autoload/go/term.vim

```diff
@@ -30,49 +30,73 @@ function! go#term#newmode(bang, cmd, errorformat, mode) abort
 
   execute l:cd . fnameescape(expand("%:p:h"))
 
-  execute l:mode . ' __go_term__'
-
-  setlocal filetype=goterm
-  setlocal bufhidden=delete
-  setlocal winfixheight
-  setlocal noswapfile
-  setlocal nobuflisted
-
-  " explicitly bind callbacks to state so that within them, self will always
-  " refer to state. See :help Partial for more information.
-  "
-  " Don't set an on_stderr, because it will be passed the same data as
-  " on_stdout. See https://github.com/neovim/neovim/issues/2836
-  let l:job = {
-        \ 'on_stdout': function('s:on_stdout', [], state),
-        \ 'on_exit' : function('s:on_exit', [], state),
-      \ }
-
-  let l:state.id = termopen(a:cmd, l:job)
-  let l:state.termwinid = win_getid(winnr())
-
-  execute l:cd . fnameescape(l:dir)
-
-  " resize new term if needed.
   let l:height = go#config#TermHeight()
   let l:width = go#config#TermWidth()
 
-  " Adjust the window width or height depending on whether it's a vertical or
-  " horizontal split.
-  if l:mode =~ "vertical" || l:mode =~ "vsplit" || l:mode =~ "vnew"
-    exe 'vertical resize ' . l:width
-  elseif mode =~ "split" || mode =~ "new"
-    exe 'resize ' . l:height
-  endif
+  " setup job for nvim
+  if has('nvim')
+    execute l:mode . ' __go_term__'
+    setlocal filetype=goterm
+    setlocal bufhidden=delete
+    setlocal winfixheight
+    setlocal noswapfile
+    setlocal nobuflisted
+
+    " explicitly bind callbacks to state so that within them, self will always
+    " refer to state. See :help Partial for more information.
+    "
+    " Don't set an on_stderr, because it will be passed the same data as
+    " on_stdout. See https://github.com/neovim/neovim/issues/2836
+    let l:job = {
+          \ 'on_stdout': function('s:on_stdout', [], state),
+          \ 'on_exit' : function('s:on_exit', [], state),
+        \ }
+    let l:state.id = termopen(a:cmd, l:job)
+    let l:state.termwinid = win_getid(winnr())
+    execute l:cd . fnameescape(l:dir)
+
+    " Adjust the window width or height depending on whether it's a vertical or
+    " horizontal split.
+    if l:mode =~ "vertical" || l:mode =~ "vsplit" || l:mode =~ "vnew"
+      exe 'vertical resize ' . l:width
+    elseif mode =~ "split" || mode =~ "new"
+      exe 'resize ' . l:height
+    endif
+    " we also need to resize the pty, so there you go...
+    call jobresize(l:state.id, l:width, l:height)
+
+  " setup term for vim8
+  elseif has('terminal')
+    let l:term = {
+          \ 'term_rows' : l:height,
+          \ 'term_cols' : l:width,
+          \ 'out_cb': function('s:out_cb', [], state),
+          \ 'exit_cb' : function('s:exit_cb', [], state),
+        \ }
+
+    if l:mode =~ "vertical" || l:mode =~ "vsplit" || l:mode =~ "vnew"
+      let l:term["vertical"] = 1
+    endif
 
-  " we also need to resize the pty, so there you go...
-  call jobresize(l:state.id, l:width, l:height)
+    let l:id = term_start(["/bin/sh", "-c", a:cmd], l:term)
```

Comment:
This won't work on Windows. Can you use `a:cmd` directly and make modify the call to `go#term#newmode` in `cmd.vim` to pass a list instead of a string?

Reply:
@bhcleek a few issues here...

vim's api will not spawn a shell "automatically" for you, instead if you pass a string, it splits it by white space, and hands it directly to execvp. When the string is split, the final argument is double quoted. The following error is encountered with GoRun

```
package '/home/louis/git/go/claircore/cmd/libindexhttp/main.go': malformed module path "'/home/louis/git/go/claircore/cmd/libindexhttp/main.go'": invalid char '\''
```
When go test is attempted and passed as a list the command fails due to no "empty" string after -tags

```
vim-go: [test -run TestSecDBParse$ -v -race -tags ] FAIL
```
Vim docs say to explicitly call your shell 

```
    The command is executed directly, not through a shell, the
                'shell' option is not used.  To use the shell: 
        let job = job_start(["/bin/sh", "-c", "echo hello"])
               Or: 
        let job = job_start('/bin/sh -c "echo hello"')
```
Should we check for unix/windows and split the logic ?

Reply:
Yes, it's executed directly, not through a shell, but the docs don't say that you _must_ use a shell. The implication from the docs seem to be that if you want a shell, then it needs to be explicitly use. Maybe I'm missing something, though?

As for `go test`. The reason the tags aren't there is because they're empty. In that case, they `-tags` shouldn't be passed at all: https://github.com/fatih/vim-go/blob/master/autoload/go/test.vim#L9 should be changed to be more like https://github.com/fatih/vim-go/blob/master/autoload/go/def.vim#L31-L34

Reply:
That is what I sought to accomplish here:
https://github.com/fatih/vim-go/pull/2639/files/e6546c8cb8947432dacec111fc2bc2c4678e0ff8#diff-c223bf93d0004ffef86b493d16e3e953R9

But you requested revert on that. Maybe my solution is not adequate here? 

Re Shell:
You are correct, shell is not required, however it passively fixed the double quoting issue which occurs here:
https://github.com/ldelossa/vim-go/blob/782297eed08ed47cedbba3b64626e532365cde61/autoload/go/cmd.vim#L126-L131

Everything works if the `go#util#Shelljoin` calls are changed to just "join" calls (vim builtins).

The issue is, the shelljoin calls are producing a string "go run 'some/path'", term_start then splits that and creates a string for each one [ "go", "run", "'some/path'"] and the double quoting of the path elements results in Go tool barking 

Let me address turning #RunTerm's cmd argument to a list first and see if it naturally solves this issue.

Reply:
Sounds good. Let me know if you need a hand.

I think all of this is a consequence of trying to use a string argument instead of a list. Once it's converted to use a list instead of a string, the quoting that Vim is generating before passing to `execvp` will stop and everything should Just Work™.